File: src/arch/x86_64/cpu/gdt.c
#include <arch/x86_64/cpu.h>

gdt_t gdt = {0};

extern void loadCS(uint16_t segment);

void initialiseGDT() {
    gdt.null = (gdt_entry_t){
        .access = 0,
        .base0 = 0,
        .base1 = 0,
        .base2 = 0,
        .limit0 = 0,
        .limit1_flags = 0
    };
    gdt.kernel_code = (gdt_entry_t){
        .access = GDT_ACCESS_ACCESSED | GDT_ACCESS_READWRITE | GDT_ACCESS_EXECUTABLE | GDT_ACCESS_DESCRIPTOR_TYPE | IDT_FLAG_DPL_0 | GDT_ACCESS_PRESENT,
        .base0 = 0,
        .base1 = 0,
        .base2 = 0,
        .limit0 = 0xFFFF,
        .limit1_flags = 0xF | GDT_FLAGS_LONG_MODE_CODE | GDT_FLAGS_GRANULARITY
    };
    gdt.kernel_data = (gdt_entry_t){
        .access = GDT_ACCESS_ACCESSED | GDT_ACCESS_READWRITE | GDT_ACCESS_DESCRIPTOR_TYPE | IDT_FLAG_DPL_0 | GDT_ACCESS_PRESENT,
        .base0 = 0,
        .base1 = 0,
        .base2 = 0,
        .limit0 = 0xFFFF,
        .limit1_flags = 0xF | GDT_FLAGS_GRANULARITY | GDT_FLAGS_SIZE
    };
    gdt.user_code = (gdt_entry_t){
        .access = GDT_ACCESS_ACCESSED | GDT_ACCESS_READWRITE | GDT_ACCESS_EXECUTABLE | GDT_ACCESS_DESCRIPTOR_TYPE | IDT_FLAG_DPL_3 | GDT_ACCESS_PRESENT,
        .base0 = 0,
        .base1 = 0,
        .base2 = 0,
        .limit0 = 0xFFFF,
        .limit1_flags = 0xF | GDT_FLAGS_LONG_MODE_CODE | GDT_FLAGS_GRANULARITY
    };
    gdt.user_data = (gdt_entry_t){
        .access = GDT_ACCESS_ACCESSED | GDT_ACCESS_READWRITE | GDT_ACCESS_DESCRIPTOR_TYPE | IDT_FLAG_DPL_3 | GDT_ACCESS_PRESENT,
        .base0 = 0,
        .base1 = 0,
        .base2 = 0,
        .limit0 = 0xFFFF,
        .limit1_flags = 0xF | GDT_FLAGS_GRANULARITY | GDT_FLAGS_SIZE
    };

    gdt_descriptor_t descriptor = {
        .base = (uint64_t)&gdt,
        .limit = sizeof(gdt_t) - 1
    };

    asm volatile ("lgdt %0" : : "m"(descriptor));

    loadCS(makeSegmentSelector(1, 0, 0));

    uint16_t data_sel = makeSegmentSelector(2, 0, 0);
    asm volatile("mov %0, %%ds" : : "r"(data_sel));
    asm volatile("mov %0, %%es" : : "r"(data_sel));
    asm volatile("mov %0, %%ss" : : "r"(data_sel));
    asm volatile("mov %0, %%fs" : : "r"(data_sel));
    asm volatile("mov %0, %%gs" : : "r"(data_sel));
}

File: src/arch/x86_64/cpu/gdt.rs
use crate::arch::x86_64::cpu::{lgdt, load_selectors};

#[repr(C, packed)]
struct Gdt {
    null: GdtEntry,
    kernel_code: GdtEntry,
    kernel_data: GdtEntry,
    user_code: GdtEntry,
    user_data: GdtEntry,
}

#[repr(C, packed)]
struct GdtEntry {
    limit_0: u16,
    base_0: u16,
    base_1: u8,
    access: u8,
    flags_limit_2: u8,
    base_2: u8,
}

#[repr(transparent)]
struct GdtAccessByte(u8);

impl GdtAccessByte {
    const fn new(
        accessed: bool,
        read_write: bool,
        direction_conforming: bool,
        executable: bool,
        descriptor_type: bool,
        dpl: u8,
        present: bool,
    ) -> GdtAccessByte {
        GdtAccessByte(
            (accessed as u8) & 0b00000001
                | ((read_write as u8) << 1) & 0b00000010
                | ((direction_conforming as u8) << 2) & 0b00000100
                | ((executable as u8) << 3) & 0b00001000
                | ((descriptor_type as u8) << 4) & 0b00010000
                | ((dpl & 0b11) << 5) & 0b01100000
                | ((present as u8) << 7) & 0b10000000,
        )
    }
}

impl GdtEntry {
    const fn new(
        base: u64,
        limit: u32,
        long_mode_code: bool,
        granularity: bool,
        size: bool,
        access: GdtAccessByte,
    ) -> GdtEntry {
        GdtEntry {
            limit_0: (limit & 0xFFFF) as u16,
            base_0: (base & 0xFFFF) as u16,
            base_1: ((base >> 16) & 0xFF) as u8,
            access: access.0,
            flags_limit_2: ((((((size as u8) << 2) & 0b0100)
                | (((granularity as u8) << 3) & 0b1000)
                | (((long_mode_code as u8) << 1) & 0b0010))
                & 0x0F)
                << 4)
                | (((limit >> 16) & 0x0F) as u8),
            base_2: ((base >> 24) & 0xFF) as u8,
        }
    }
}

#[repr(C, packed)]
pub struct GdtPointer {
    limit: u16,
    base: u64,
}

static GDT: Gdt = Gdt {
    null: GdtEntry::new(
        0,
        0,
        false,
        false,
        false,
        GdtAccessByte::new(false, false, false, false, false, 0, false),
    ),
    kernel_code: GdtEntry::new(
        0,
        0xFFFFFFFF,
        true,
        true,
        false,
        GdtAccessByte::new(true, true, false, true, true, 0, true),
    ),
    kernel_data: GdtEntry::new(
        0,
        0xFFFFFFFF,
        false,
        true,
        true,
        GdtAccessByte::new(true, true, false, false, true, 0, true),
    ),
    user_code: GdtEntry::new(
        0,
        0xFFFFFFFF,
        true,
        true,
        false,
        GdtAccessByte::new(true, true, false, true, true, 3, true),
    ),
    user_data: GdtEntry::new(
        0,
        0xFFFFFFFF,
        false,
        true,
        true,
        GdtAccessByte::new(true, true, false, false, true, 3, true),
    ),
};

pub fn init() {
    let descriptor = GdtPointer {
        limit: core::mem::size_of::<Gdt>() as u16 - 1,
        base: &GDT as *const _ as u64,
    };

    lgdt(&descriptor);

    load_selectors();
}
